File: 2024/hj/week7/PacificAtlanticWaterFlow.cpp
class Solution {
public:
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        int m = heights.size();
        int n = heights[0].size();

        vector<vector<int>> result;
        vector<vector<bool>> pacific(m, vector<bool>(n, false));
        vector<vector<bool>> atlantic(m, vector<bool>(n, false));

        for (int i = 0; i < m; i++) {
            dfs(i, 0, pacific, heights);
            dfs(i, n - 1, atlantic, heights);
        }
        for (int j = 0; j < n; j++) {
            dfs(0, j, pacific, heights);
            dfs(m - 1, j, atlantic, heights);
        }

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (pacific[i][j] && atlantic[i][j]) {
                    result.push_back({i, j});
                }
            }
        }

        return result;
    }

private:
    void dfs(int r, int c, vector<vector<bool>>& ocean, vector<vector<int>>& heights) {
        ocean[r][c] = true;

        vector<pair<int, int>> directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        int m = heights.size();
        int n = heights[0].size();

        for (auto [dr, dc] : directions) {
            int nr = r + dr;
            int nc = c + dc;

            if (nr >= 0 && nr < m &&
                nc >= 0 && nc < n &&
                !ocean[nr][nc] &&
                heights[nr][nc] >= heights[r][c]) {
                dfs(nr, nc, ocean, heights);
            }
        }
    }
};

File: 2025/Jun/pacific_atlantic_water_flow.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        row_len = len(heights)
        col_len = len(heights[0])
        pacific = set()
        atlantic = set()
        def dfs(visited, row, col, prev_height):
            if 0 <= row < row_len and 0 <= col < col_len and heights[row][col] >= prev_height and (row, col) not in visited:
                visited.add((row, col))
                current_height = heights[row][col]
                dfs(visited, row+1, col, current_height)
                dfs(visited, row-1, col, current_height)
                dfs(visited, row, col+1, current_height)
                dfs(visited, row, col-1, current_height)

        res = []
        for i in range(row_len):
            dfs(pacific, i, 0, heights[i][0])
            dfs(atlantic, i, col_len - 1, heights[i][col_len - 1])

        for i in range(col_len):
            dfs(pacific, 0, i, heights[0][i])
            dfs(atlantic, row_len - 1, i, heights[row_len - 1][i])

        for (r,c) in pacific & atlantic:
            res.append([r,c])

        return res

